File: scripts/dev_tools/codex_native_converter/reporting.py
"""Write review and apply artifacts for the Codex-native converter.

Purpose:
    Emit the required report artifact set for review and apply runs in a stable,
    deterministic layout.

Usage:
    The converter engine calls ``write_conversion_report_set`` after validation
    so every run produces a report, mapping catalog, validation results, and a
    reviewable proposed tree.

Flow:
    Structured data is rendered first, artifact directories are created next,
    then the Markdown report, JSON catalogs, and proposed-tree files are
    written in deterministic path order.

Invariants / Constraints:
    Artifact filenames and serialization ordering remain stable across runs for
    the same inputs.

Side Effects:
    Writes report artifacts through the configured filesystem adapter.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import TYPE_CHECKING, Protocol

if TYPE_CHECKING:
    from pathlib import Path

    from scripts.dev_tools.codex_native_converter.models import (
        MappingRecord,
        RunOptions,
        TopologyEdge,
        TranslationTrace,
        ValidationFinding,
    )

# Import topology rendering helpers from the dedicated topology module.
from scripts.dev_tools.codex_native_converter._reporting_topology import (
    render_destination_to_repeated_source_chart,
    render_source_to_destination_chart,
    render_source_to_repeated_destination_chart,
)


class ConverterFileSystem(Protocol):
    """Describe the filesystem operations needed by report writing.

    Purpose:
        Keep report generation testable without requiring temporary filesystem
        writes in the unit tests.

    Usage:
        Production code passes ``RealConverterFileSystem`` and tests may pass a
        fake implementation.

    Flow:
        Callers create directories first and then write report files through
        this interface.

    Invariants / Constraints:
        Implementations must preserve the provided path and text content.

    Side Effects:
        Implementation dependent.
    """

    def mkdir(self, path: Path) -> None:
        """Create a directory and any missing parent directories."""

    def write_text(self, path: Path, content: str) -> None:
        """Write UTF-8 text content to one path."""


class RealConverterFileSystem:
    """Write converter output to the real filesystem.

    Purpose:
        Provide the default report-writing adapter for production CLI runs.

    Usage:
        Passed into ``write_conversion_report_set`` by default.

    Flow:
        Directory creation ensures parent folders exist before text is written.

    Invariants / Constraints:
        Paths are written exactly as provided.

    Side Effects:
        Creates directories and writes files on disk.
    """

    def mkdir(self, path: Path) -> None:
        """Create a directory and any missing parents.

        Purpose:
            Ensure one artifact directory exists before files are written.

        Args:
            path (Path): Directory path to create.

        Returns:
            None: This method returns no value.

        Raises:
            OSError: Raised when the directory cannot be created.

        Side Effects:
            Creates directories on disk.
        """

        path.mkdir(parents=True, exist_ok=True)

    def write_text(self, path: Path, content: str) -> None:
        """Write UTF-8 text content to one path.

        Purpose:
            Persist a report artifact or proposed output file.

        Args:
            path (Path): File path to write.
            content (str): UTF-8 text content to persist.

        Returns:
            None: This method returns no value.

        Raises:
            OSError: Raised when the file cannot be written.

        Side Effects:
            Writes a text file on disk.
        """

        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(content, encoding="utf-8")


@dataclass(frozen=True, slots=True)
class ReportSetPaths:
    """Describe the artifact paths written for one converter run.

    Purpose:
        Return the artifact index to engine callers and automation wrappers.

    Usage:
        The engine includes this value in review/apply summaries and TypeScript
        wrappers can surface the Markdown report path to users.

    Flow:
        The report writer constructs the paths before writing the artifact set.

    Invariants / Constraints:
        All stored paths are absolute paths rooted beneath the artifact root.

    Side Effects:
        None.
    """

    conversion_report: Path
    mapping_catalog: Path
    validation_results: Path
    proposed_tree_root: Path


def _render_conversion_report(
    run_options: RunOptions,
    mapping_records: tuple[MappingRecord, ...],
    topology_edges: tuple[TopologyEdge, ...],
    translation_traces: tuple[TranslationTrace, ...],
    validation_findings: tuple[ValidationFinding, ...],
) -> str:
    """Render the human-readable Markdown conversion report.

    Purpose:
        Provide a concise artifact that reviewers can read without opening the
        JSON catalogs directly.

    Args:
        run_options (RunOptions): Requested run options.
        mapping_records (tuple[MappingRecord, ...]): Planned mappings.
        validation_findings (tuple[ValidationFinding, ...]): Validation results.

    Returns:
        str: Markdown report text.

    Raises:
        None.

    Side Effects:
        None.
    """

    blocking_count = sum(1 for finding in validation_findings if finding.blocking)
    destination_root_text = (
        run_options.destination_root.as_posix()
        if run_options.destination_root is not None
        else "review-only"
    )
    sorted_mapping_records = tuple(
        sorted(mapping_records, key=lambda record: record.source_path)
    )
    sorted_topology_edges = tuple(
        sorted(
            topology_edges,
            key=lambda edge: (edge.source_path, edge.destination_path),
        )
    )
    sorted_translation_traces = tuple(
        sorted(
            translation_traces,
            key=lambda trace: (
                trace.source_path,
                trace.section_id,
                trace.target_role.value,
                trace.target_path or "",
            ),
        )
    )
    lines = [
        "# Conversion Report",
        "",
        f"- Mode: `{run_options.mode}`",
        f"- Source ecosystem: `{run_options.source_ecosystem.value}`",
        f"- Source root: `{run_options.source_root.as_posix()}`",
        f"- Destination root: `{destination_root_text}`",
        f"- Artifact root: `{run_options.artifact_root.as_posix()}`",
        f"- Mapping records: {len(mapping_records)}",
        (
            f"- Validation findings: {len(validation_findings)} "
            f"({blocking_count} blocking)"
        ),
        "",
        "## Mapping Topology",
        "",
        "### Shared Destination Nodes",
        "",
        "Source and destination nodes are both deduplicated in this view.",
        "",
    ]
    lines.extend(render_source_to_destination_chart(sorted_topology_edges))
    lines.extend(
        (
            "",
            "### Repeated Destination Nodes",
            "",
            (
                "Destination nodes may repeat in this source-to-destination "
                "view so fan-in stays legible."
            ),
        )
    )
    lines.extend(render_source_to_repeated_destination_chart(sorted_topology_edges))
    lines.extend(
        (
            "",
            "### Repeated Source Nodes",
            "",
            (
                "Source nodes may repeat in this destination-to-source view so "
                "fan-out stays legible."
            ),
        )
    )
    lines.extend(render_destination_to_repeated_source_chart(sorted_topology_edges))
    lines.extend(
        (
            "",
            "## Mappings",
            "",
            "| Source path | Conversion class | Target role | Target path | Notes |",
            "| --- | --- | --- | --- | --- |",
        )
    )

    # Render mappings in stable source-path order so review diffs stay small
    # and predictable.
    for mapping_record in sorted_mapping_records:
        notes = "<br>".join(mapping_record.notes) if mapping_record.notes else ""
        lines.append(
            "| "
            f"`{mapping_record.source_path}` | "
            f"`{mapping_record.conversion_class.value}` | "
            f"`{mapping_record.target_role.value}` | "
            f"`{mapping_record.target_path or ''}` | {notes} |"
        )

    lines.extend(["", "## Section Mappings", ""])
    if not sorted_translation_traces:
        lines.append("- None")
    else:
        lines.extend(
            (
                (
                    "| Source path | Section | Intent | Target role | "
                    "Target path | Notes |"
                ),
                "| --- | --- | --- | --- | --- | --- |",
            )
        )
        for translation_trace in sorted_translation_traces:
            notes = (
                "<br>".join(translation_trace.notes) if translation_trace.notes else ""
            )
            lines.append(
                "| "
                f"`{translation_trace.source_path}` | "
                f"`{translation_trace.heading}` | "
                f"`{translation_trace.intent_kind.value}` | "
                f"`{translation_trace.target_role.value}` | "
                f"`{translation_trace.target_path or ''}` | "
                f"{notes} |"
            )

    lines.extend(["", "## Validation Findings", ""])
    if not validation_findings:
        lines.append("- None")
    else:
        # Render validation findings in a stable order so the Markdown summary
        # mirrors JSON output.
        for validation_finding in sorted(
            validation_findings,
            key=lambda finding: (
                finding.code,
                finding.source_path or "",
                finding.target_path or "",
            ),
        ):
            lines.append(f"- `{validation_finding.code}`: {validation_finding.message}")

    return "\n".join(lines) + "\n"


def write_conversion_report_set(
    run_options: RunOptions,
    mapping_records: tuple[MappingRecord, ...],
    topology_edges: tuple[TopologyEdge, ...],
    translation_traces: tuple[TranslationTrace, ...],
    validation_findings: tuple[ValidationFinding, ...],
    generated_output: dict[str, str],
    *,
    fs: ConverterFileSystem | None = None,
) -> ReportSetPaths:
    """Write the required review/apply report artifact set.

    Purpose:
        Emit the required Markdown report, JSON catalogs, and proposed-tree
        snapshot for one converter run.

    Args:
        run_options (RunOptions): Requested run options.
        mapping_records (tuple[MappingRecord, ...]): Planned mappings.
        topology_edges (tuple[TopologyEdge, ...]): Derived topology edges for
            Mermaid report rendering.
        translation_traces (tuple[TranslationTrace, ...]): Section-level
            translation traces for mixed-content report rendering.
        validation_findings (tuple[ValidationFinding, ...]): Validation results.
        generated_output (dict[str, str]): Generated output keyed by target path.
        fs (ConverterFileSystem | None): Filesystem adapter for writes.

    Returns:
        ReportSetPaths: Paths to the written report artifacts.

    Raises:
        OSError: Raised when an artifact cannot be written.

    Side Effects:
        Creates artifact directories and writes the report set through the
        configured filesystem adapter.
    """

    resolved_fs = fs or RealConverterFileSystem()
    artifact_root = run_options.artifact_root.resolve()
    proposed_tree_root = artifact_root / "proposed-tree"
    report_paths = ReportSetPaths(
        conversion_report=artifact_root / "conversion-report.md",
        mapping_catalog=artifact_root / "mapping-catalog.json",
        validation_results=artifact_root / "validation-results.json",
        proposed_tree_root=proposed_tree_root,
    )

    resolved_fs.mkdir(artifact_root)
    resolved_fs.mkdir(proposed_tree_root)

    mapping_catalog_payload = [
        mapping_record.to_json_dict()
        for mapping_record in sorted(
            mapping_records, key=lambda record: record.source_path
        )
    ]
    validation_results_payload = [
        validation_finding.to_json_dict()
        for validation_finding in sorted(
            validation_findings,
            key=lambda finding: (
                finding.code,
                finding.source_path or "",
                finding.target_path or "",
            ),
        )
    ]

    resolved_fs.write_text(
        report_paths.conversion_report,
        _render_conversion_report(
            run_options,
            mapping_records,
            topology_edges,
            translation_traces,
            validation_findings,
        ),
    )
    resolved_fs.write_text(
        report_paths.mapping_catalog,
        json.dumps(mapping_catalog_payload, indent=2, sort_keys=True) + "\n",
    )
    resolved_fs.write_text(
        report_paths.validation_results,
        json.dumps(validation_results_payload, indent=2, sort_keys=True) + "\n",
    )

    # Write the proposed tree in stable path order so review runs always emit a
    # deterministic snapshot of the generated content.
    for target_path in sorted(generated_output):
        resolved_fs.write_text(
            proposed_tree_root / target_path, generated_output[target_path]
        )

    return report_paths

File: extensions/drm-copilot/src/lib/codex-native-converter/reporting.ts
/**
 * Write review and apply artifacts for the Codex-native converter.
 *
 * Purpose:
 *     Emit the required report artifact set for review and apply runs in a
 *     stable, deterministic layout. Ported from `reporting.py`; the Markdown
 *     report renderer lives in `reporting-render.ts` so neither file exceeds the
 *     500-line policy. Writes flow through the injected {@link FileSystem}.
 *
 * Invariants:
 *     Artifact filenames and serialization ordering remain stable across runs
 *     for the same inputs. JSON catalogs use sorted keys, 2-space indentation,
 *     and a trailing newline, matching the Python `json.dumps` usage.
 */

import { type FileSystem } from "../file-system";
import {
  type MappingRecord,
  mappingRecordToJson,
  type RunOptions,
  type TopologyEdge,
  type TranslationTrace,
  type ValidationFinding,
  validationFindingToJson,
} from "./models";
import { renderConversionReport } from "./reporting-render";

/**
 * The artifact paths written for one converter run.
 *
 * Mirrors `ReportSetPaths`. All stored paths are rooted beneath the artifact
 * root.
 */
export interface ReportSetPaths {
  readonly conversionReport: string;
  readonly mappingCatalog: string;
  readonly validationResults: string;
  readonly proposedTreeRoot: string;
}

/**
 * Recursively sort object keys so serialization matches Python `sort_keys=True`.
 *
 * @param value Value to normalize for deterministic serialization.
 * @returns A structurally equivalent value with all object keys sorted.
 */
function sortKeysDeep(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map((item) => sortKeysDeep(item));
  }
  if (value !== null && typeof value === "object") {
    const sorted: Record<string, unknown> = {};
    for (const key of Object.keys(value as Record<string, unknown>).sort(
      (left, right) => (left < right ? -1 : left > right ? 1 : 0),
    )) {
      sorted[key] = sortKeysDeep((value as Record<string, unknown>)[key]);
    }
    return sorted;
  }
  return value;
}

/**
 * Serialize a JSON-safe payload with sorted keys, 2-space indentation, and a
 * trailing newline, matching `json.dumps(..., indent=2, sort_keys=True) + "\n"`.
 *
 * @param payload JSON-safe value to serialize.
 * @returns The deterministic JSON string with a trailing newline.
 */
function dumpsSortedWithNewline(payload: unknown): string {
  return JSON.stringify(sortKeysDeep(payload), null, 2) + "\n";
}

/**
 * Collapse `.`/`..` segments in a POSIX path, mirroring `Path.resolve` segment
 * handling for already-absolute inputs.
 *
 * @param value POSIX path.
 * @returns The path with relative segments collapsed.
 */
function collapseSegments(value: string): string {
  const isAbsolute = value.startsWith("/");
  const driveMatch = /^([A-Za-z]:)(.*)$/.exec(value);
  const drive = driveMatch ? (driveMatch[1] ?? "") : "";
  const body = driveMatch ? (driveMatch[2] ?? "") : value;
  const resolved: string[] = [];
  for (const segment of body.split("/")) {
    if (segment === "" || segment === ".") {
      continue;
    }
    if (segment === "..") {
      if (resolved.length > 0) {
        resolved.pop();
      }
      continue;
    }
    resolved.push(segment);
  }
  const joined = resolved.join("/");
  if (drive) {
    return `${drive}/${joined}`;
  }
  return isAbsolute ? `/${joined}` : joined;
}

/**
 * Join a POSIX root with a child path segment.
 *
 * @param root POSIX root path.
 * @param child Child path segment.
 * @returns The combined POSIX path.
 */
function joinPosix(root: string, child: string): string {
  const normalizedRoot = root.replace(/\/+$/, "");
  const normalizedChild = child.replace(/^\/+/, "");
  return normalizedRoot === ""
    ? normalizedChild
    : `${normalizedRoot}/${normalizedChild}`;
}

/**
 * Sort mapping records by source path, returning a new array.
 *
 * @param mappingRecords Mapping records to sort.
 * @returns The records sorted by source path.
 */
function sortBySourcePath(
  mappingRecords: ReadonlyArray<MappingRecord>,
): MappingRecord[] {
  return [...mappingRecords].sort((left, right) =>
    left.sourcePath < right.sourcePath
      ? -1
      : left.sourcePath > right.sourcePath
        ? 1
        : 0,
  );
}

/**
 * Sort validation findings by (code, sourcePath, targetPath).
 *
 * @param validationFindings Findings to sort.
 * @returns The findings sorted by stable key.
 */
function sortFindings(
  validationFindings: ReadonlyArray<ValidationFinding>,
): ValidationFinding[] {
  const compare = (left: string, right: string): number =>
    left < right ? -1 : left > right ? 1 : 0;
  return [...validationFindings].sort((left, right) => {
    const byCode = compare(left.code, right.code);
    if (byCode !== 0) {
      return byCode;
    }
    const bySource = compare(left.sourcePath ?? "", right.sourcePath ?? "");
    if (bySource !== 0) {
      return bySource;
    }
    return compare(left.targetPath ?? "", right.targetPath ?? "");
  });
}

/**
 * Write the required review/apply report artifact set.
 *
 * Mirrors `write_conversion_report_set`: resolves the artifact root, creates
 * the artifact and proposed-tree directories, writes the Markdown report and
 * sorted JSON catalogs, and writes the proposed tree in sorted path order.
 *
 * @param fileSystem Injected filesystem.
 * @param runOptions Requested run options.
 * @param mappingRecords Planned mappings.
 * @param topologyEdges Derived topology edges for Mermaid rendering.
 * @param translationTraces Section-level translation traces.
 * @param validationFindings Validation results.
 * @param generatedOutput Generated output keyed by target path.
 * @returns Paths to the written report artifacts.
 * @throws Error When an artifact cannot be written.
 */
export function writeConversionReportSet(
  fileSystem: FileSystem,
  runOptions: RunOptions,
  mappingRecords: ReadonlyArray<MappingRecord>,
  topologyEdges: ReadonlyArray<TopologyEdge>,
  translationTraces: ReadonlyArray<TranslationTrace>,
  validationFindings: ReadonlyArray<ValidationFinding>,
  generatedOutput: Readonly<Record<string, string>>,
): ReportSetPaths {
  const artifactRoot = collapseSegments(runOptions.artifactRoot);
  const proposedTreeRoot = joinPosix(artifactRoot, "proposed-tree");
  const reportPaths: ReportSetPaths = {
    conversionReport: joinPosix(artifactRoot, "conversion-report.md"),
    mappingCatalog: joinPosix(artifactRoot, "mapping-catalog.json"),
    validationResults: joinPosix(artifactRoot, "validation-results.json"),
    proposedTreeRoot,
  };

  fileSystem.ensureDir(artifactRoot);
  fileSystem.ensureDir(proposedTreeRoot);

  const mappingCatalogPayload = sortBySourcePath(mappingRecords).map((record) =>
    mappingRecordToJson(record),
  );
  const validationResultsPayload = sortFindings(validationFindings).map(
    (finding) => validationFindingToJson(finding),
  );

  fileSystem.writeTextFile(
    reportPaths.conversionReport,
    renderConversionReport(
      runOptions,
      mappingRecords,
      topologyEdges,
      translationTraces,
      validationFindings,
    ),
  );
  fileSystem.writeTextFile(
    reportPaths.mappingCatalog,
    dumpsSortedWithNewline(mappingCatalogPayload),
  );
  fileSystem.writeTextFile(
    reportPaths.validationResults,
    dumpsSortedWithNewline(validationResultsPayload),
  );

  // Write the proposed tree in stable path order so review runs always emit a
  // deterministic snapshot of the generated content.
  for (const targetPath of Object.keys(generatedOutput).sort((left, right) =>
    left < right ? -1 : left > right ? 1 : 0,
  )) {
    fileSystem.writeTextFile(
      joinPosix(proposedTreeRoot, targetPath),
      generatedOutput[targetPath] ?? "",
    );
  }

  return reportPaths;
}
